pyiron_base/jobs/job/generic.py

```diff
@@ -1174,14 +1174,14 @@ def _to_dict(self):
             data_dict["files_to_compress"] = self._files_to_remove
         return data_dict
 
-    def from_dict(self, job_dict):
-        self._type_from_dict(type_dict=job_dict)
-        if "import_directory" in job_dict.keys():
-            self._import_directory = job_dict["import_directory"]
-        self._server.from_dict(server_dict=job_dict["server"])
-        if "executable" in job_dict.keys() and job_dict["executable"] is not None:
-            self.executable.from_dict(job_dict["executable"])
-        input_dict = job_dict["input"]
+    def _from_dict(self, obj_dict, version=None):
+        self._type_from_dict(type_dict=obj_dict)
+        if "import_directory" in obj_dict.keys():
+            self._import_directory = obj_dict["import_directory"]
+        self._server = obj_dict["server"]
+        if "executable" in obj_dict.keys() and obj_dict["executable"] is not None:
+            self._executable = obj_dict["executable"]
```

Comment:
I am a bit surprised about this part, when is the dictionary of the executable converted to the executable object? Is this already happening when reading the executable from the HDF5 file?

Reply:
In this new setup HasDict.from_dict converts nested objects back from their dictionary form before the implementations HasDict._from_dict is called.  So by the time GenericJob._from_dict runs and looks for the executable Executable._from_dict already ran.

Reply:
In principle the reverse also works, i.e. it's not always necessary anymore to call to_dict on ones children that are returned in the obj_dict.

Reply:
I think it makes sense to document this at least in the `HasDict` class, so developers understand how to derive their own classes from `HasDict` and that they only have to implement the `_to_dict()` and `_from_dict()` interface. Finally, I am wondering if it makes sense to overload the `__getstate__()` and `__setstate__()` function in the `HasDict` class. Then we can slowly transition from calling `to_dict()` and `from_dict()` to using the `__getstate__()` and `__setstate__()` interface to reload the pyiron objects. This would simplify the interface for developers who just want to integrate a new code. As long as their classes can be pickled, then we can store their classes in our jobs. If they want to optimise the performance and benefit from the hierarchical nature of the HDF5 file, then they can overload the `__getstate__()` and `__setstate__()` method, for example by attaching a dataclass to their class which they use for data storage or by attaching a data container.